### ISSUE TITLE
substreams-sink-mongodb causes the substreams to freeze in production mode

### ISSUE THREAD

**Issue by sduchesneau** [streamingfast/substreams#156]
Steps to reproduce:

1) run a mongodb from substreams-sink-mongodb/devel/eth-block-meta/mongodb (`docker-compose up`)
2) from your browser (localhost:8081), ensure that you delete the 'substreams' table so you don't have data in your local mongo
3) on the substreams server side, delete the two store caches:
`c5984fb9d09c6588146d69b47933167826b52ea1` and `8c5c3880a2813e633634edbfd56dd9df3319a623` 
4) with you substreams token correctly set in your environment variable, run `devel/eth-block-meta/start.sh` 
watch as the last_block stays #10011 forever:
```
2023-01-24T10:03:03.147-0500 INFO (sink-mongodb) session init {"trace_id": "eddccfd58fa7f564874c1bfaf46eb367"}
2023-01-24T10:03:05.022-0500 INFO (sink-mongodb) substreams sink stats {"progress_msg_rate": "0.000 msg/s (0 total)", "block_rate": "364.000 blocks/s (364 total)", "last_block": "#1017 (5ac49cf828da73b56ca9ba190dc2cb43800a9488ed41e8d2c0ced5cf3a51918c)"}
2023-01-24T10:03:07.023-0500 INFO (sink-mongodb) substreams sink stats {"progress_msg_rate": "0.000 msg/s (0 total)", "block_rate": "545.000 blocks/s (1635 total)", "last_block": "#2330 (bbfbee11c8df48bef490a294b3447dfbee96cae39b511adcd1f86331a039ec32)"}
2023-01-24T10:03:09.022-0500 INFO (sink-mongodb) substreams sink stats {"progress_msg_rate": "60.000 msg/s (300 total)", "block_rate": "631.400 blocks/s (3157 total)", "last_block": "#3970 (20753b24dd98c97735d9acdca7dabaaf1fd420f221ece287850106c1c8440d1e)"}
2023-01-24T10:03:11.022-0500 INFO (sink-mongodb) substreams sink stats {"progress_msg_rate": "296.750 msg/s (2374 total)", "block_rate": "670.143 blocks/s (4691 total)", "last_block": "#5425 (582ab60870590bb5f274d8f7ac117635249331573c81db1b428a30e852c758d3)"}
2023-01-24T10:03:13.022-0500 INFO (sink-mongodb) substreams sink stats {"progress_msg_rate": "396.000 msg/s (5505 total)", "block_rate": "690.000 blocks/s (6210 total)", "last_block": "#6966 (c28febd2bd021bc76bc763b5b988e2ca0c978cbf93eda534d49b02b40d6bcb6c)"}
2023-01-24T10:03:15.022-0500 INFO (sink-mongodb) substreams sink stats {"progress_msg_rate": "638.364 msg/s (7022 total)", "block_rate": "701.273 blocks/s (7714 total)", "last_block": "#8323 (2d63232a81b10e991e9b78a9eb2a4abd9f3786a6621dae39ef5b10d7f0209ca9)"}
2023-01-24T10:03:17.022-0500 INFO (sink-mongodb) substreams sink stats {"progress_msg_rate": "795.929 msg/s (11143 total)", "block_rate": "693.923 blocks/s (9021 total)", "last_block": "#9781 (2eb419e13a8b574b4abf9c694882c35d612cfd351f96cc7dce6848b1779e2334)"}
2023-01-24T10:03:19.026-0500 INFO (sink-mongodb) substreams sink stats {"progress_msg_rate": "814.667 msg/s (12220 total)", "block_rate": "667.467 blocks/s (10012 total)", "last_block": "#10011 (e64db02ff8543ee1f284b0cc01d7b82799d59972d7bf2010b8d4e8b26448fc61)"}
2023-01-24T10:03:21.023-0500 INFO (sink-mongodb) substreams sink stats {"progress_msg_rate": "718.941 msg/s (12222 total)", "block_rate": "588.941 blocks/s (10012 total)", "last_block": "#10011 (e64db02ff8543ee1f284b0cc01d7b82799d59972d7bf2010b8d4e8b26448fc61)"}
2023-01-24T10:03:23.022-0500 INFO (sink-mongodb) substreams sink stats {"progress_msg_rate": "643.368 msg/s (12224 total)", "block_rate": "526.947 blocks/s (10012 total)", "last_block": "#10011 (e64db02ff8543ee1f284b0cc01d7b82799d59972d7bf2010b8d4e8b26448fc61)"}
...
```

**Comment by sduchesneau** [comment]
found that it happens only when using TrafficDirector (XDS).
May be caused by a bad handling of context...

**Comment by sduchesneau** [comment]
tier1 is stuck trying to read message from tier2

```
goroutine 1808 [select, 4 minutes]:
google.golang.org/grpc/internal/transport.(*recvBufferReader).readClient(0xc001235090, {0xc01eedcdf0, 0x5, 0x5})
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.50.1/internal/transport/transport.go:191 +0xa5
google.golang.org/grpc/internal/transport.(*recvBufferReader).Read(0xc001235090, {0xc01eedcdf0?, 0xc01eb742e8?, 0xc001a12ba8?})
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.50.1/internal/transport/transport.go:171 +0x147
google.golang.org/grpc/internal/transport.(*transportReader).Read(0xc00ff283c0, {0xc01eedcdf0?, 0xc001a12c20?, 0xba7c07?})
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.50.1/internal/transport/transport.go:485 +0x32
io.ReadAtLeast({0x3e9cc00, 0xc00ff283c0}, {0xc01eedcdf0, 0x5, 0x5}, 0x5)
	/opt/hostedtoolcache/go/1.18.10/x64/src/io/io.go:331 +0x9a
io.ReadFull(...)
	/opt/hostedtoolcache/go/1.18.10/x64/src/io/io.go:350
google.golang.org/grpc/internal/transport.(*Stream).Read(0xc000383560, {0xc01eedcdf0, 0x5, 0x5})
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.50.1/internal/transport/transport.go:469 +0xa5
google.golang.org/grpc.(*parser).recvMsg(0xc01eedcde0, 0x40000000)
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.50.1/rpc_util.go:559 +0x47
google.golang.org/grpc.recvAndDecompress(0x33a1c60?, 0xc000383560, {0x0, 0x0}, 0x40000000, 0x0, {0x0, 0x0})
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.50.1/rpc_util.go:690 +0x66
google.golang.org/grpc.recv(0x3ec8bd8?, {0x7f2304598a58, 0x5d9abd8}, 0x0?, {0x0?, 0x0?}, {0x332af20, 0xc00e5de440}, 0xc009eb7b00?, 0x0, ...)
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.50.1/rpc_util.go:756 +0x6e
google.golang.org/grpc.(*csAttempt).recvMsg(0xc00182ab40, {0x332af20?, 0xc00e5de440}, 0xc001bbef40?)
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.50.1/stream.go:1025 +0x2b0
google.golang.org/grpc.(*clientStream).RecvMsg.func1(0x0?)
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.50.1/stream.go:865 +0x25
google.golang.org/grpc.(*clientStream).withRetry(0xc001299d40, 0xc001a13070, 0xc001a13040)
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.50.1/stream.go:708 +0x3be
google.golang.org/grpc.(*clientStream).RecvMsg(0xc001299d40, {0x332af20?, 0xc00e5de440?})
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.50.1/stream.go:864 +0x12e
go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc.(*clientStream).RecvMsg(0xc008e10cc0, {0x332af20, 0xc00e5de440})
	/home/runner/go/pkg/mod/go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc@v0.36.4/interceptor.go:150 +0x42
github.com/streamingfast/substreams/pb/sf/substreams/v1.(*streamBlocksClient).Recv(0xc01eb70680)
	/home/runner/go/pkg/mod/github.com/streamingfast/substreams@v0.1.1-0.20230124203635-c525fae98e57/pb/sf/substreams/v1/substreams_grpc.pb.go:62 +0x4c
github.com/streamingfast/substreams/orchestrator/work.(*RemoteWorker).Work(0xc00e1d3ec0, {0x3ec10f0?, 0xc0068c1ef0?}, 0xc00182a6c0, 0xc0002d69c0)
	/home/runner/go/pkg/mod/github.com/streamingfast/substreams@v0.1.1-0.20230124203635-c525fae98e57/orchestrator/work/worker.go:149 +0x127e
github.com/streamingfast/substreams/orchestrator.(*Scheduler).runSingleJob.func1({0x3ec10f0?, 0xc0068c1ef0?})
	/home/runner/go/pkg/mod/github.com/streamingfast/substreams@v0.1.1-0.20230124203635-c525fae98e57/orchestrator/scheduler.go:190 +0x8b
github.com/streamingfast/derr.RetryContext.func1({0x3ec10f0?, 0xc0068c1ef0?})
	/home/runner/go/pkg/mod/github.com/streamingfast/derr@v0.0.0-20221125175206-82e01d420d45/retry.go:24 +0x2b
github.com/sethvargo/go-retry.Do({0x3ec10f0, 0xc0068c1ef0}, {0x3ea8a00, 0xc0000c2a20}, 0xc001a13c70)
	/home/runner/go/pkg/mod/github.com/sethvargo/go-retry@v0.2.3/retry.go:60 +0x98
github.com/streamingfast/derr.RetryContext({0x3ec10f0, 0xc0068c1ef0}, 0x3225656c62617473?, 0xc0015f7e40)
	/home/runner/go/pkg/mod/github.com/streamingfast/derr@v0.0.0-20221125175206-82e01d420d45/retry.go:23 +0x65
github.com/streamingfast/substreams/orchestrator.(*Scheduler).runSingleJob(0xc0015992c0, {0x3ec10f0, 0xc0068c1ef0}, {0x3eae0b8, 0xc00e1d3ec0}, 0xc0085799c0, 0xc0011e3e00)
	/home/runner/go/pkg/mod/github.com/streamingfast/substreams@v0.1.1-0.20230124203635-c525fae98e57/orchestrator/scheduler.go:185 +0x20e
github.com/streamingfast/substreams/orchestrator.(*Scheduler).run.func1()
	/home/runner/go/pkg/mod/github.com/streamingfast/substreams@v0.1.1-0.20230124203635-c525fae98e57/orchestrator/scheduler.go:104 +0x8a
created by github.com/streamingfast/substreams/orchestrator.(*Scheduler).run
	/home/runner/go/pkg/mod/github.com/streamingfast/substreams@v0.1.1-0.20230124203635-c525fae98e57/orchestrator/scheduler.go:103 +0x465
```

**Comment by sduchesneau** [comment]
while the tier2 worker filling the store stopped way before it was over and is stuck here:
```goroutine 185 [select, 11 minutes]:
google.golang.org/grpc/internal/transport.(*writeQuota).get(0xc0012e8020, 0x19)
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.50.1/internal/transport/flowcontrol.go:59 +0x6e
google.golang.org/grpc/internal/transport.(*http2Server).Write(0xc00051c9c0, 0xc001a29200, {0xc0033b38d8, 0x5, 0x5}, {0xc003251050, 0x14, 0x14}, 0x0?)
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.50.1/internal/transport/http2_server.go:1093 +0x16a
google.golang.org/grpc.(*serverStream).SendMsg(0xc0017402a0, {0x332af20?, 0xc00063ef00})
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.50.1/stream.go:1578 +0x1c2
github.com/grpc-ecosystem/go-grpc-prometheus.(*monitoredServerStream).SendMsg(0xc001713c98, {0x332af20?, 0xc00063ef00?})
	/home/runner/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-prometheus@v1.2.0/server_metrics.go:156 +0x33
go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc.(*serverStream).SendMsg(0xc001f0fef0, {0x332af20, 0xc00063ef00})
	/home/runner/go/pkg/mod/go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc@v0.36.4/interceptor.go:382 +0x42
go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc.(*serverStream).SendMsg(0xc001f1e0f0, {0x332af20, 0xc00063ef00})
	/home/runner/go/pkg/mod/go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc@v0.36.4/interceptor.go:382 +0x42
github.com/streamingfast/substreams/pb/sf/substreams/v1.(*streamBlocksServer).Send(0x203000?, 0x203000?)
	/home/runner/go/pkg/mod/github.com/streamingfast/substreams@v0.1.1-0.20230124203635-c525fae98e57/pb/sf/substreams/v1/substreams_grpc.pb.go:112 +0x2b
github.com/streamingfast/substreams/service.responseHandler.func1(0x332ae20?)
	/home/runner/go/pkg/mod/github.com/streamingfast/substreams@v0.1.1-0.20230124203635-c525fae98e57/service/service.go:342 +0x43
github.com/streamingfast/substreams/tracking.(*bytesMeter).Send(0xc001783da0, {0x3ec10f0?, 0xc001bf4030?}, 0xc0013e4300)
	/home/runner/go/pkg/mod/github.com/streamingfast/substreams@v0.1.1-0.20230124203635-c525fae98e57/tracking/bytes.go:123 +0x2e5
github.com/streamingfast/substreams/tracking.(*bytesMeter).Start(0xc00013afc0?, {0x3ec10f0, 0xc001bf4030}, 0xc0014797a0?)
	/home/runner/go/pkg/mod/github.com/streamingfast/substreams@v0.1.1-0.20230124203635-c525fae98e57/tracking/bytes.go:76 +0xdc
created by github.com/streamingfast/substreams/tracking.(*bytesMeter).Launch
	/home/runner/go/pkg/mod/github.com/streamingfast/substreams@v0.1.1-0.20230124203635-c525fae98e57/tracking/bytes.go:85 +0xb0
```

**Comment by sduchesneau** [comment]
disabling the bytes tracker solves the issue.
It is weird because (IIRC), the tier2 is sending other kind of messages than the 'bytes' one.. 
this solved our issue for now, we'll need to do more digging before we can reactivate the bytes counter message

**Comment by maoueh** [comment]
Closing, let's track the root cause in another issue
